File: source/lib/state_representation/cpp/src/Robot/JacobianMatrix.cpp
#include "state_representation/Robot/JacobianMatrix.hpp"
#include "state_representation/Exceptions/IncompatibleStatesException.hpp"
#include "state_representation/Exceptions/EmptyStateException.hpp"

namespace StateRepresentation 
{
	JacobianMatrix::JacobianMatrix(const std::string& robot_name, unsigned int nb_joints):
	State("JacobianMatrix", robot_name), nb_rows(6), nb_cols(nb_joints)
	{
		this->set_joint_names(nb_joints);
	}

	JacobianMatrix::JacobianMatrix(const std::string& robot_name, const std::vector<std::string>& joint_names):
	State("JacobianMatrix", robot_name), nb_rows(6), nb_cols(joint_names.size())
	{
		this->set_joint_names(joint_names);
	}

	JacobianMatrix::JacobianMatrix(const std::string& robot_name, const Eigen::MatrixXd& data):
	State("JacobianMatrix", robot_name), nb_rows(data.rows()), nb_cols(data.cols())
	{
		this->set_joint_names(this->get_nb_cols());
		this->set_data(data);
	}

	JacobianMatrix::JacobianMatrix(const std::string& robot_name, const std::vector<std::string>& joint_names, const Eigen::MatrixXd& data):
	State("JacobianMatrix", robot_name), nb_rows(data.rows()), nb_cols(joint_names.size())
	{
		this->set_joint_names(joint_names);
		this->set_data(data);
	}

	JacobianMatrix::JacobianMatrix(const JacobianMatrix& jacobian):
	State(jacobian), joint_names(jacobian.joint_names), data(jacobian.data)
	{}

	void JacobianMatrix::initialize()
	{
		this->State::initialize();
		this->data.resize(this->nb_rows, this->nb_cols);
		this->data.setZero();
	}

	const JacobianMatrix JacobianMatrix::transpose()
	{
		JacobianMatrix result(*this);
		result.set_nb_cols(this->get_nb_rows());
		result.set_nb_rows(this->get_nb_cols());
		result.set_data(result.get_data().transpose());
		return result;
	}

	const Eigen::MatrixXd JacobianMatrix::operator*(const Eigen::MatrixXd& matrix) const
	{
		if(this->is_empty()) throw EmptyStateException(this->get_name() + " state is empty");
		if(matrix.rows() != this->get_nb_cols()) throw IncompatibleSizeException("Input matrix is of incorrect size");
		return this->get_data() * matrix;
	}

	const CartesianTwist JacobianMatrix::operator*(const JointVelocities& dq) const
	{
		if(dq.is_empty()) throw EmptyStateException(dq.get_name() + " state is empty");
		if(!this->is_compatible(dq)) throw IncompatibleStatesException("The JacobianMatrix and the input JointVelocities are incompatible"o);
		Eigen::Matrix<double, 6, 1> twist = (*this) * dq.get_velocities();
		// crate a CartesianTwist with name "robot"_end_effector and reference frame "robot"_base
		CartesianTwist result(this->get_name() + "_end_effector", twist, this->get_name() + "_base");
		return result;
	}

	const Eigen::MatrixXd JacobianMatrix::solve(const Eigen::MatrixXd& matrix) const
	{
		if(this->is_empty()) throw EmptyStateException(this->get_name() + " state is empty");
		if(this->get_nb_rows() != matrix.rows()) throw IncompatibleSizeException("Input matrix is of incorrect size");
		return this->get_data().colPivHouseholderQr().solve(matrix);
	}

	const JointVelocities JacobianMatrix::solve(const CartesianTwist& dX) const
	{
		if(dX.is_empty()) throw EmptyStateException(dX.get_name() + " state is empty");
		// extract the velocity as a 6d vector
		auto twist = dX.get_twist();
		// this use the solve operation instead of using the inverse or pseudo-inverse of the jacobian
		Eigen::VectorXd joint_velocities = (*this).solve(twist);
		// return a JointVelocities state
		JointVelocities result(this->get_name(), this->get_joint_names(), joint_velocities);
		return result;
	}

	const JointVelocities JacobianMatrix::operator*(const CartesianTwist& dX) const
	{
		return this->solve(dX);
	}

	const JacobianMatrix JacobianMatrix::copy() const
	{
		JacobianMatrix result(*this);
		return result;
	}

	std::ostream& operator<<(std::ostream& os, const JacobianMatrix& matrix)
	{
		if(matrix.is_empty())
		{
			os << "Empty JacobianMatrix";
		}
		else
		{
			os << matrix.get_name() << " JacobianMatrix" << std::endl;
			os << "joint names: [";
			for(auto& n:matrix.get_joint_names()) os << n << ", ";
			os << "]" << std::endl;
			os << "number of rows: " << matrix.get_nb_rows() << std::endl;
			os << "number of columns: " << matrix.get_nb_cols() << std::endl;
			for(unsigned int i = 0; i < matrix.get_nb_rows(); ++i)
			{
				os << "| " << matrix(i,0);
				for(unsigned int j = 1; j < matrix.get_nb_cols(); ++j)
				{
					os << ", " << matrix(i,j);
				}
				os << " |" << std::endl;
			}
		}
		return os;
	}
}

File: source/lib/state_representation/python/src/state_representation/robot/jacobian_matrix.py
#!/Ussr/bin/python
from state_representation.state import State

import numpy as np
import math
import copy
import warnings


class JacobianMatrix(state):
    def __init__(self, *args, robot_name=None, nb_joints=None, joint_names=None, data=None,
                 jacobian=None):
        
        if len(args): # Nonzero
            if isinstance(args[0], JacobianMatrix):
                jacobian=args[0]
            elif isinstance(args[0], str):
                robot_name = args[0]
                if len(args)>1:
                    for ii in range(1, len(args)):
                        if isinstance(args[ii], int):
                            nb_joints = args[ii]
                        elif (isinstance(args[ii], list) and isinstance(args[ii][0], str)):
                            joint_names = args[ii]
                        elif isinstance(args[ii], (list, np.ndarray)):
                            data = args[ii]
                        else:
                            raise TypeError()
            else:
                raise TypeError()
            
        # Choose type
        if isinstance(jacobian, JacobianMatrix):
            super().__init__(jacobian)
            self.joint_names = jacobian.joint_names
            self.joint_data = jacobian.joint_data
            
        elif isinstance(robot_name, str):
            super().__init__(robot_name, str)

            if isinstance(joint_names, list):
                self.joint_names = joint_names
            elif:
                isinstance(nb_joints, int):
                self.joint_names(nb_joints)
                
            if isinstance(data, list):
                self.data = data

                self.nb_rows = self._data.shape[0]
                self.nb_cols = self._data.shape[1]
        else:
            raise ValueError("No input defined")

    # def initialize(self):
        # super().initialize()
        # self.data = self.data.reshape(nb_rows, )

    @property
    def nb_rows(self):
        return self._nb_rows

    @nb_rows.setter
    def nb_rows(self):
        return self._nb_rows
    
    @property
    def nb_cols(self):
        return self._nb_cols

    @nb_cols.setter
    def nb_cols(self):
        return self._nb_cols

    @data.setter
    def data(self, value):
        value = np.squeeze(value)
        if len(value.shape) == 2:
            self._data = value
        else:
            raise TypeError()

    @property
    def data(self, value):
        return self._data
        
    @joint_names.setter
    def joint_names(self, value):
        if isinstance(value, int):
            self._joint_names = ["joint"+ii for ii in range(value)]
        elif isinstance(value, list):
            self._joint_names
        else:
            TypeError()

    @property
    def joint_names(self):
        return self._joint_names
    
    def transpose(self):
        result = copy.deepcopy(self)
        result.data = self.data.T
        
    @property
    def T():
        return self.transpose()

    def initialize():
        self._is_empty = True
        
    def is_compabible(state):
        return (self.name==state.name and self.reference_frame==state.reference_frame)

    def __imul__(self, other):
        if self.is_empty():
                raise RuntimeError()
        if isinstance(other, JointVelocities):
            if other.is_empty():
                raise RuntimeError()
            
            if not self.is_compabible(other):
                raise ValueError("The JacobianMatrix and the input JointVelocities are incompatible")
            twist = self * other.velocity
            return CartesianTwist(this.name+"_end_effector", twist, self.name+"base")
        elif isinstance(other, (list, np.array)):
            other = np.squeeze(other)

        elif isinstance(other, CartesianTwist):
            return self.solve(other)
        else:
            raise TypeError()
        
    def __mul__(self, other):
        result = copy.deepcopy(self)
        result *= other
        return result


    def solve(self. other):
        if self.is_empty():
            raise RuntimeError()

        if isinstance(other, np.ndarray):
            if self.nb_rows != other.shape[0]:
                raise ValueError("Input matrix is of incorrect size")
            return self.data
            # return this->get_data().colPivHouseholderQr().solve(matrix)
        
            
        elif isinstance(other, CartesianTwist):
            if other.is_empty():
                raise RuntimeError()
            twist = other.get_twist()
            joint_velocities = this.solve(twist)

            return JointVelocities(self.name, self.joint_names, joint_velocities)
            
        elif isinstance(other, list):
            return self.solve(np.sqeeze(ohter))
        
        else:
            raise TypeError()

    def copy(self, other=None):
        if isintance(other, type(None)):
            return copy.deepcopy(self)
        else:
            return copy.deepcopy(other)

    def __str__(self):
        if self.is_empty:
            res = "Empty JacobianMatrix"
        else:
            res = self.name + " JacobianMatrix expressed in " + self.reference_frame + "\n"
            res += "joint names: [" + " ".join([str(x) for x in self.joint_names]) + "]\n"
            res += "number of rows: " + self.nb_rows""
            res += "number of cols: " + self.nb_cols""
            res += "data: "
            res += np.array2string(np.self.data) # TODO - check this
        return res
        
        

    
